### DISCUSSION TITLE
New project, new pyenv venv, eglot times out to pyright-langserver

### DISCUSSION THREAD

**Discussion by Cerebus** [joaotavora/eglot#1166] [Q&A] [answered]
Not sure what's going on here as I get no stderr to look at, but:

- New venv with pyenv, pyright installed along with my other standard set of packages.
- pyenv local set.
- eglot-ensure set.

I can see eglot spin up pyright; it just times out talking to it, kills the server, and exits.  From messages:

```
[jsonrpc] Server exited with status 9
Error running timer: (error "[eglot] Timed out after 30 seconds")
```

If I swap out the project venv for another project's venv, everything's fine.  This leads me to think it's something about the venv, but damnfino what.

Looking at eglot's events buffer, there's nothing unusual in the startup:

- The server binary is the correct one for the venv.
- rootPath and rootUri are correct.

Any ideas what to poke at?

**Comment by Cerebus** [comment, marked as answer]
Version splitting, breaks as of pyright 1.1.291.

**Reply by ndryden** [reply]
Hi! Were you able to identify a root cause for this other than the version change? I am also seeing this issue, but only on some platforms. (It breaks on Linux, but works macOS, for example.)

**Reply by joaotavora** [reply]
I have a Linux box with a Archlinux and system-installed Python 3.  If you provide precise fail-proof instructions on how to install this language server and reproduce this problem, I can help analyze it. Be sure to run these instructions yourself before sharing them, so I don't waste time troubleshooting toochain or emacs init file issues needlessly.

**Reply by ndryden** [reply]
I hope this is sufficient, please let me know if there's other details. I'm using a completely fresh install of emacs 28.2 and eglot from ELPA. No other packages or customization.

Install Pyright in a new venv, set up a dummy Python file:
```
$ python3 -m venv $HOME/venvs/test
$ source $HOME/venvs/test/bin/activate
$ python3 -m pip install pyright
$ pyright --version
pyright 1.1.293
$ mkdir test
$ touch test/test.py
```

Open `test/test.py` with emacs, `M-x eglot`. It should detect Pyright and start it, then Pyright is killed 30 seconds later:
```
Error running timer: (error "[eglot] Timed out after 30 seconds")
```

Eglot events buffer:
```
[internal] Fri Feb 10 08:28:16 2023:
(:message "Running language server: /home/ndryden/venvs/test/bin/pyright-langserver --stdio")
[client-request] (id:1) Fri Feb 10 08:28:16 2023:
(:jsonrpc "2.0" :id 1 :method "initialize" :params
          (:processId 48607 :rootPath "/home/ndryden/test/" :rootUri "file:///home/ndryden/test" :initializationOptions #s(hash-table size 1 test eql rehash-size 1.5 rehash-threshold 0.8125 data
                                                                                                                                        ())
                      :capabilities
                      (:workspace
                       (:applyEdit t :executeCommand
                                   (:dynamicRegistration :json-false)
                                   :workspaceEdit
                                   (:documentChanges t)
                                   :didChangeWatchedFiles
                                   (:dynamicRegistration t)
                                   :symbol
                                   (:dynamicRegistration :json-false)
                                   :configuration t :workspaceFolders t)
                       :textDocument
                       (:synchronization
                        (:dynamicRegistration :json-false :willSave t :willSaveWaitUntil t :didSave t)
                        :completion
                        (:dynamicRegistration :json-false :completionItem
                                              (:snippetSupport :json-false :deprecatedSupport t :resolveSupport
                                                               (:properties
                                                                ["documentation" "details" "additionalTextEdits"])
                                                               :tagSupport
                                                               (:valueSet
                                                                [1]))
                                              :contextSupport t)
                        :hover
                        (:dynamicRegistration :json-false :contentFormat
                                              ["plaintext"])
                        :signatureHelp
                        (:dynamicRegistration :json-false :signatureInformation
                                              (:parameterInformation
                                               (:labelOffsetSupport t)
                                               :activeParameterSupport t))
                        :references
                        (:dynamicRegistration :json-false)
                        :definition
                        (:dynamicRegistration :json-false :linkSupport t)
                        :declaration
                        (:dynamicRegistration :json-false :linkSupport t)
                        :implementation
                        (:dynamicRegistration :json-false :linkSupport t)
                        :typeDefinition
                        (:dynamicRegistration :json-false :linkSupport t)
                        :documentSymbol
                        (:dynamicRegistration :json-false :hierarchicalDocumentSymbolSupport t :symbolKind
                                              (:valueSet
                                               [1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26]))
                        :documentHighlight
                        (:dynamicRegistration :json-false)
                        :codeAction
                        (:dynamicRegistration :json-false :codeActionLiteralSupport
                                              (:codeActionKind
                                               (:valueSet
                                                ["quickfix" "refactor" "refactor.extract" "refactor.inline" "refactor.rewrite" "source" "source.organizeImports"]))
                                              :isPreferredSupport t)
                        :formatting
                        (:dynamicRegistration :json-false)
                        :rangeFormatting
                        (:dynamicRegistration :json-false)
                        :rename
                        (:dynamicRegistration :json-false)
                        :publishDiagnostics
                        (:relatedInformation :json-false :codeDescriptionSupport :json-false :tagSupport
                                             (:valueSet
                                              [1 2])))
                       :experimental #s(hash-table size 1 test eql rehash-size 1.5 rehash-threshold 0.8125 data
                                                   ()))
                      :workspaceFolders
                      [(:uri "file:///home/ndryden/test" :name "~/test/")]))
[internal] Fri Feb 10 08:28:46 2023:
(:message "Connection state changed" :change "killed\n")

----------b---y---e---b---y---e----------
[stderr]
[stderr]
[stderr] nil
[stderr] nil
[stderr] Process EGLOT (test/(python-mode python-ts-mode)) stderr finished
```

**Reply by joaotavora** [reply]
> I hope this is sufficient, please let me know if there's other details. I

This seems enough.  I'll try to look at this this weekend

**Reply by joaotavora** [reply]
In the meantime, can you try to compile the latest Emacs `master` (or the 29.1 pretest) or otherwise try with the `jsonrpc.el` library that lives there?  This behaviour _could_ be caused by a problem that I fixed recently (in that `jsonrpc.el` library) and which might not have made it to the ELPA repository somehow.

**Reply by ndryden** [reply]
I just tested with the latest `master` and I see the same behavior as above.

**Reply by joaotavora** [reply]
I've just reproduced the problem. It seems that this new version of Pyright is mixing the stderr and stdout in the same channel.  If I type simply:

```
pyright 2> /dev/null
```

to get rid of stderr (i.e. redirect it to /dev/null) I should _not_ get any output back from the program. But with this `pyright`, I do.  Can you try to do the same with an older version of pyright?

For comparison, `pylsp` behaves correctly. Just issuing `pylsp` gets you some stderr visible output, but `pylsp 2> /dev/null` is silent.

stdout is reserved for JSONRPC messages, so if you put the usual debugging junk that goes on stderr there, JSONRPC communication is affected, and here it break entirely, as Emacs is waiting for a JSONRPC response and sees some goofy stuff before that response.  Maybe other JSONRPC clients can deal with this, but `jsonrpc.el` can't and I haven't yet encountered a LSP server in the many tens of servers that have been tested that does this.

I'd suggest investigating what changed on the pyright side and perhaps reporting this to them.

**Reply by ndryden** [reply]
Thanks for digging into this!

I believe I've traced the issue to a recent change in the Python wrapper package for Pyright and have opened an issue and suggested fix there (RobertCraigie/pyright-python#149).

Edit: This should be fixed now and presumably will show up in the next release.
